### PR TITLE
ARC

### PR DESCRIPTION
**Pull request by vbvg2008**

### CODE REVIEW

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 1]
```diff
@@ -37,38 +43,176 @@ class LRScheduler(Trace):
...
+        train_lr = np.array(train_lr) / train_lr[-1]
+        train_lr = cv2.resize(train_lr, (1, 300), interpolation=cv2.INTER_NEAREST)
+        return train_lr
+
+    def _preprocess_train_loss(self, train_loss: list, missing: int) -> np.ndarray:
+        target_size = (3 - missing) * 100
+        train_loss = np.array(train_loss, dtype="float32")
+        train_loss = cv2.resize(train_loss, (1, target_size))
+        train_loss = self._zscore(train_loss)
+        if train_loss.size < 300:
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:198]: we should fix this to actually be correct rather than continuing an accidental mistake

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 2]
```diff
@@ -37,38 +43,176 @@ class LRScheduler(Trace):
...
+            train_loss = np.pad(train_loss, ((300 - train_loss.size, 0), (0, 0)), mode='constant', constant_values=0.0)
+        return train_loss
+
+    def _zscore(self, data: np.ndarray, epsilon=1e-7):
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:202]: this should probably be a backend function

File: fastestimator/trace/adapt/lr_scheduler.py (hunk 1 above)
Comment (vbvg2008) [reply fastestimator/trace/adapt/lr_scheduler.py:198]: the model is trained by the accidental mistake.     When we improve ARC, we will fix this.

Comment (TortoiseHam) [reply fastestimator/trace/adapt/lr_scheduler.py:198]: it may actually perform better even without retraining if we fix this

Comment (vbvg2008) [reply fastestimator/trace/adapt/lr_scheduler.py:198]: i know, not now :)

Comment (TortoiseHam) [reply fastestimator/trace/adapt/lr_scheduler.py:198]: it's not like we're under immediate time pressure to put this specific version in

Comment (vbvg2008) [reply fastestimator/trace/adapt/lr_scheduler.py:198]: if we retrain the model, then we have to redo the entire evaluation on multiple tasks.    The initial version should be consistent with the one we used in paper.

The ARC research will continue and we will update the model later.

Comment (TortoiseHam) [reply fastestimator/trace/adapt/lr_scheduler.py:198]: the research code will be published separately regardless. by the time the paper is out we'll already have moved on here with improvements anyways

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 3]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
     s = LRScheduler(model=model, lr_fn=lambda epoch: fe.schedule.cosine_decay(epoch, cycle_length=3750, init_lr=1e-3))
     fe.Estimator(..., traces=[s])  # Learning rate will change based on epoch
+    s = LRScheduler(model=model, lr_fn='arc')
+    fe.Estimator(..., traces=[s])  # Learning rate will change based on epoch using ARC algorithm
     ```
 
     Args:
         model: A model instance compiled with fe.build.
-        lr_fn: A lr scheduling function that takes either 'epoch' or 'step' as input.
+        lr_fn: A lr scheduling function that takes either 'epoch' or 'step' as input or string 'arc'.
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:53]:
```suggestion
        lr_fn: A lr scheduling function that takes either 'epoch' or 'step' as input, or the string 'arc'.
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 4]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        frequency: The frequency of applying the lr_fn, in integer.
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:54]:
```suggestion
        frequency: The frequency with which to invoke the `lr_fn`, or None for default.
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 5]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        if self.system.mode == "train" and self.schedule_mode == "epoch":
+            if isinstance(self.lr_fn, ARC):
+                if (self.system.epoch_idx % self.frequency == 1 or self.frequency == 1) and self.system.epoch_idx > 1:
+                    multiplier = self.lr_fn.predict_next_multiplier()
+                    new_lr = np.float32(get_lr(model=self.model) * multiplier)
+                    set_lr(self.model, new_lr)
+                    print("FastEstimator-ARC: Multiplying LR by {}".format(multiplier))
+            elif self.system.epoch_idx % self.frequency == 0:
+                new_lr = np.float32(self.lr_fn(self.system.epoch_idx))
+                set_lr(self.model, new_lr)
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:99]:
```suggestion
        if self.system.mode == "train" and self.schedule_mode == "epoch" and (self.system.epoch_idx % self.frequency == 1 or self.frequency == 1):
            if isinstance(self.lr_fn, ARC):
                if self.system.epoch_idx > 1:
                    multiplier = self.lr_fn.predict_next_multiplier()
                    new_lr = np.float32(get_lr(model=self.model) * multiplier)
                    set_lr(self.model, new_lr)
                    print("FastEstimator-ARC: Multiplying LR by {}".format(multiplier))
            else:
                new_lr = np.float32(self.lr_fn(self.system.epoch_idx))
                set_lr(self.model, new_lr)
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 6]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
 
     def on_batch_begin(self, data: Data) -> None:
-        if self.schedule_mode == "step":
+        if self.system.mode == "train" and self.schedule_mode == "step" and \
+            self.system.global_step % self.frequency == 0:
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:103]:
```suggestion
            (self.system.global_step % self.frequency == 1 or self.frequency == 1):
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 7]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+            if self.system.epoch_idx % self.frequency == 0:
+                self.lr_fn.gather_multiple_eval_losses()
+        if self.system.mode == "train" and isinstance(self.lr_fn, ARC) and self.system.epoch_idx % self.frequency == 0:
+            self.lr_fn.accumulate_all_lrs(get_lr(model=self.model))
+            self.lr_fn.gather_multiple_train_losses()
+
+
+@traceable()
+class ARC:
+    def __init__(self, weights_path) -> None:
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:127]:
```suggestion
    def __init__(self, weights_path: str) -> None:
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 8]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        if self.system.mode == "train" and self.schedule_mode == "step" and \
+            self.system.global_step % self.frequency == 0:
             new_lr = np.float32(self.lr_fn(self.system.global_step))
             set_lr(self.model, new_lr)
 
     def on_batch_end(self, data: Data) -> None:
-        if self.system.log_steps and (self.system.global_step % self.system.log_steps == 0
-                                      or self.system.global_step == 1):
+        if self.system.mode == "train" and isinstance(self.lr_fn, ARC):
+            self.lr_fn.accumulate_single_train_loss(data[next(iter(self.model.loss_name))].numpy())
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:109]:
```suggestion
            self.lr_fn.accumulate_single_train_loss(data[min(self.model.loss_name)].numpy())
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 9]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        if self.system.mode == "train" and self.system.log_steps and (
+                self.system.global_step % self.system.log_steps == 0 or self.system.global_step == 1):
             current_lr = np.float32(get_lr(self.model))
             data.write_with_log(self.outputs[0], current_lr)
+
+    def on_epoch_end(self, data: Data) -> None:
+        if self.system.mode == "eval" and isinstance(self.lr_fn, ARC):
+            self.lr_fn.accumulate_single_eval_loss(data[next(iter(self.model.loss_name))])
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:117]:
```suggestion
            self.lr_fn.accumulate_single_eval_loss(data[min(self.model.loss_name)])
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 10]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+    def _preprocess_train_loss(self, train_loss: list, missing: int) -> np.ndarray:
+        target_size = (3 - missing) * 100
+        train_loss = np.array(train_loss, dtype="float32")
+        train_loss = cv2.resize(train_loss, (1, target_size))
+        train_loss = zscore(train_loss)
+        if train_loss.size < 300:
+            train_loss = np.pad(train_loss, ((300 - train_loss.size, 0), (0, 0)), mode='constant', constant_values=0.0)
+        return train_loss
+
+    def _merge_list(self, data: list):
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:203]:
```suggestion
    def _merge_list(self, data: List[Union[None, float, List[float]]) -> Tuple[List[float], int]:
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 11]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        train_lr, _ = self._merge_list(self.all_train_lr)
+        train_lr = self._preprocess_train_lr(train_lr)
+        model_inputs = np.concatenate((train_loss, val_loss, train_lr), axis=1)
+        model_inputs = np.expand_dims(model_inputs, axis=0)
+        with tf.device("cpu:0"):
+            model_pred = feed_forward(model=self.model, x=model_inputs, training=False)
+        action = np.argmax(model_pred)
+        return self.lr_multiplier[action]
+
+    def _preprocess_val_loss(self, val_loss: list) -> np.ndarray:
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:181]:
```suggestion
    def _preprocess_val_loss(self, val_loss: List[float]) -> np.ndarray:
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 12]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        if val_loss:
+            val_loss = zscore(np.array(val_loss))
+            val_loss = cv2.resize(val_loss, (1, 300), interpolation=cv2.INTER_NEAREST)
+        else:
+            val_loss = np.zeros([300, 1], dtype="float32")
+        return val_loss
+
+    def _preprocess_train_lr(self, train_lr: list) -> np.ndarray:
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:189]:
```suggestion
    def _preprocess_train_lr(self, train_lr: List[float]) -> np.ndarray:
```

File: fastestimator/trace/adapt/lr_scheduler.py [hunk 13]
```diff
@@ -37,38 +44,170 @@ class LRScheduler(Trace):
...
+        train_lr = np.array(train_lr) / train_lr[-1]
+        train_lr = cv2.resize(train_lr, (1, 300), interpolation=cv2.INTER_NEAREST)
+        return train_lr
+
+    def _preprocess_train_loss(self, train_loss: list, missing: int) -> np.ndarray:
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/lr_scheduler.py:194]:
```suggestion
    def _preprocess_train_loss(self, train_loss: List[float], missing: int) -> np.ndarray:
```

File: test/PR_test/integration_test/trace/adapt/test_lr_scheduler.py [hunk 14]
```diff
@@ -16,6 +16,14 @@
 import unittest
 
 import fastestimator as fe
+import numpy as np
```
Comment (TortoiseHam) [inline test/PR_test/integration_test/trace/adapt/test_lr_scheduler.py:19]: import sort doesn't look right here
